iroha_p2p/src/network.rs

```diff
@@ -93,222 +113,264 @@ where
+            Ok(peer) => peer.start().await,
+            Err(e) => {
+                warn!(%e, "Unable to create peer");
+                return;
             }
-            Err(e) => warn!(%e, "Unable to create peer"),
-        }
+        };
     }
 }
 
 #[async_trait::async_trait]
 impl<T, K, E> Handler<Post<T>> for NetworkBase<T, K, E>
 where
-    T: Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
+    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
     type Result = ();
 
     async fn handle(&mut self, msg: Post<T>) {
-        let addr: &Addr<Peer<T, K, E>> = &self.peers[&msg.id];
-        addr.do_send(msg).await;
+        match self.peers.get(&msg.id.public_key) {
+            Some(conn) => conn.addr.do_send(msg).await,
+            None if msg.id.public_key == self.public_key => debug!("Not sending message to myself"),
+            None => info!(
+                "Didn't find peer to send message, have only {} connections!",
+                self.peers.len()
+            ),
+        }
     }
 }
 
 #[async_trait::async_trait]
-impl<T, K, E> Handler<PeerMessage<T>> for NetworkBase<T, K, E>
+impl<T, K, E> Handler<GotConnected<T, K, E>> for NetworkBase<T, K, E>
 where
-    T: Encode + Decode + BrokerMessage + Send + Sync + Clone + 'static,
+    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
     type Result = ();
 
-    async fn handle(&mut self, msg: PeerMessage<T>) {
-        match msg {
-            PeerMessage::Connected(id, connection_id) => {
-                if self.connected_peers.contains(&id.public_key) {
-                    warn!(
-                        "Peer with public key {:?} already connected!",
-                        &id.public_key
-                    );
-                    self.broker.issue_send(StopSelf(connection_id)).await;
-                } else {
-                    let _ = self.connected_peers.insert(id.public_key);
-                }
-            }
-            PeerMessage::Disconnected(id) => {
-                self.peers.remove(&id);
-                let _ = self.connected_peers.remove(&id.public_key);
+    async fn handle(&mut self, GotConnected { id, conn }: GotConnected<T, K, E>) {
+        debug!(?id, "Connected");
+        let addr = &self.listen_addr;
+        match self.peers.entry(id.public_key.clone()) {
+            Entry::Occupied(mut entry) => {
+                warn!(?addr, ?id, "Twin connection");
+                warn!(
+                    ?addr,
+                    out = conn.outgoing,
+                    keys = self.public_key > id.public_key
+                );
+
+                // If we initiated connection and our public key is bigger, then keep new connection
+                let addr = match (conn.outgoing, self.public_key > id.public_key) {
```

Comment:
This will not work. See my PR.

Reply:
Sure, how can it be done then?

Reply:
I've solved it like this:
1. Hold a vector of connection to any public key: `peers: HashMap<PublicKey, Vec<Connection<T, K, E>>>`.
2. In `Sumeragi.connect_peers()` I get a list of already connected peers and initiate connections only to those not connected.

The main difficulty of solving twin connections is the fact that those 2 connections can be processed in different order on their nodes. Like 1,2 and 2,1. Then, the code that is checking for twins is run in different order too.

Reply:
Also, I have removed connection attempts loop. Peer tries to connect just once. If it can't, the Sumeragi will retry it in a second.

Reply:
So, how do you check for twins then? Like in the case of 1,2 and 2,1?